src/relay/analysis/extract_operators.cc

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file extract_operators.cc
+ * \brief Extract unique operators from an IRModule
+ */
+#include <tvm/node/structural_hash.h>
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/expr.h>
+#include <tvm/relay/expr_functor.h>
+
+namespace tvm {
+namespace relay {
+
+class OperatorExtractorWrapper : private ExprVisitor {
+ public:
+  explicit OperatorExtractorWrapper(const IRModule& mod) : mod_(mod) {}
+
+  Array<String> Extract() {
+    VisitExpr(this->mod_->Lookup("main"));
+
+    return this->operators;
+  }
+
+ private:
+  const IRModule mod_;
+  // Array of unique operator names.
+  Array<String> operators;
+
+  void VisitExpr_(const OpNode* n) final {
```

Comment:
IMHO, it would be more useful if this pass can also report how many times each op is used in this model. Moreover, this pass can also report fused ops by visiting FunctionNode with kPrimitive=1.

Reply:
Yeah we could start to generalize this for "static" extraction if we also extracted use counts as well.

Reply:
That's a good point. I'm actually not against the current implementation but just did brain-storming. It's fine if this PR would like to deliver a list of unique op names first. I'll let @anwang2009 decide.

Reply:
Changed to {op name: freq} extraction

Reply:
For fused ops, the pass traverses the fused function to count the operators -- is that what you had in mind? @comaniac 

```
#[version = "0.0.5"]
def @main(%x: Tensor[(1, 1, 5, 1), float32], %w1: Tensor[(1, 1, 3, 3), float32], %w2: Tensor[(1, 1, 3, 3), float32]) -> Tensor[(1, 1, 5, 1), float32] {
  %1 = fn (%p01: Tensor[(1, 1, 5, 1), float32], %p11: Tensor[(1, 1, 3, 3), float32], Primitive=1) -> Tensor[(1, 1, 5, 1), float32] {
    nn.conv2d(%p01, %p11, padding=[1, 1, 1, 1], channels=1, kernel_size=[3, 3]) /* ty=Tensor[(1, 1, 5, 1), float32] */
  };
  %2 = %1(%x, %w1) /* ty=Tensor[(1, 1, 5, 1), float32] */;
  %3 = fn (%p0: Tensor[(1, 1, 5, 1), float32], %p1: Tensor[(1, 1, 3, 3), float32], Primitive=1) -> Tensor[(1, 1, 5, 1), float32] {
    %0 = nn.conv2d(%p0, %p1, padding=[1, 1, 1, 1], channels=1, kernel_size=[3, 3]) /* ty=Tensor[(1, 1, 5, 1), float32] */;
    add(%p0, %0) /* ty=Tensor[(1, 1, 5, 1), float32] */
  };
  %3(%2, %w2) /* ty=Tensor[(1, 1, 5, 1), float32] */
}
```
Extraction of the above mod gives `{"add": 1, "nn.conv2d": 2}`

Reply:
Yeah I've considered this before. I think it's up to you and both are fine as long as we make it clear in the docstring. If needed, we can support both and let users decide (something like `ignore_fused_ops: True`).

Reply:
ok, I just added a comment. Not supporting both for now.